NExtends/Collections/ObjectPropertiesList.cs

```diff
@@ -0,0 +1,362 @@
+	///		i[1] = "test0";
+	///		i[2] = "test1";
+	///		i[3] = "test2";
+	///
+	///		Assert.IsTrue(t.test == "test");
+	///		Assert.IsTrue(t.field0 == "test0");
+	///		Assert.IsTrue(t.field1 == "test1");
+	///		Assert.IsTrue(t.field2 == "test2");
+	///	}
+	///}
+	///
+	///[ObjectPropertiesListProperties(0, "test")]
+	///[ObjectPropertiesListNumberedProperties(1, "field", 3, 0)]
+	///class Test
+	///{
+	///		public string test { get; set; }
+	///		public string field0 { get; set; }
+	///		public string field1 { get; set; }
+	///		public string field2 { get; set; }
+	///}
+	///</code>
+	///</example>
+	public class ObjectPropertiesList<K, T> : IList<T>
+	{
+		public static IEnumerable<string> IndexableProperties;
+		public static Func<K, T>[] Getters;
+		public static Action<K, T>[] Setters;
+		public static string[] PropertyNames;
+
+		K ObjectToIndex;
+
+		static ObjectPropertiesList()
+		{
+			var getters = new List<Func<K, T>>();
+			var setters = new List<Action<K, T>>();
+			var propertyNames = new List<string>();
+
+			var Attributes = typeof(K).GetTypeInfo().GetCustomAttributes(typeof(IObjectPropertiesListParams), true).OfType<IObjectPropertiesListParams>();
+
+			if (Attributes.Count() == 0)
+			{
+				//in case we don't have any specified attributes we use all public properties
+				IndexableProperties = typeof(K).GetProperties().Where(p => p.PropertyType == typeof(T)).Select(p => p.Name).ToList();
+			}
+			else
+			{
+				IndexableProperties = Attributes.OrderBy(a => a.Order).SelectMany(a => a.IndexableProperties).ToList();
+			}
+
+			foreach (var pName in IndexableProperties)
+			{
+				var pInfo = typeof(K).GetProperty(pName);
+
+				if (pInfo == null)
+					throw new Exception(string.Format("Class {0} doesn't have a property named {1}", typeof(K).Name, pName));
+
+				var getter = pInfo.GetValueGetter<K, T>();
+				var setter = pInfo.GetValueSetter<K, T>();
+
+				getters.Add(getter);
+				setters.Add(setter);
+				propertyNames.Add(pName);
+			}
+
+			Getters = getters.ToArray();
+			Setters = setters.ToArray();
+			PropertyNames = propertyNames.ToArray();
+		}
+
+		public ObjectPropertiesList(K objectToIndex)
+		{
+			this.ObjectToIndex = objectToIndex;
+		}
+
+		public T this[int key]
+		{
+			get
+			{
+				return Getters[key](ObjectToIndex);
+			}
+			set
+			{
+				Setters[key](ObjectToIndex, value);
+			}
+		}
+
+		public void Add(T item)
+		{
+			throw new NotImplementedException();
+		}
+
+		public void Clear()
+		{
+			throw new NotImplementedException();
+		}
+
+		public bool Contains(T item)
+		{
+			bool contains = false;
+			for (int i = 0; i < Count && !contains; i++)
+				contains = this[i].Equals(item);
+
+			return contains;
+		}
+
+		public void CopyTo(T[] array, int arrayIndex)
+		{
+			if (object.ReferenceEquals(array, null))
+			{
+				throw new ArgumentNullException(
+					"Null array reference",
+					"array"
+					);
+			}
+
+			if (arrayIndex < 0)
+			{
+				throw new ArgumentOutOfRangeException(
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') The parameter name 'Index is out of range' is not declared in the argument list. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3928)